File: JobFinder-Android/backend/src/main/java/com/deepakmanktala/jobfinder/repository/JobRepository.java
package com.deepakmanktala.jobfinder.repository;

import com.deepakmanktala.jobfinder.model.Job;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

public interface JobRepository extends JpaRepository<Job, Long> {

    @Query("""
        SELECT j FROM Job j
        WHERE (:keyword IS NULL OR LOWER(j.title) LIKE LOWER(CONCAT('%', :keyword, '%')))
          AND (:region IS NULL OR LOWER(j.region) LIKE LOWER(CONCAT('%', :region, '%')))
          AND (:source IS NULL OR LOWER(j.source) LIKE LOWER(CONCAT('%', :source, '%')))
        """)
    Page<Job> searchJobs(@Param("keyword") String keyword,
                         @Param("region") String region,
                         @Param("source") String source,
                         Pageable pageable);
}

File: JobFinder-Android/android/app/src/main/java/com/deepakmanktala/jobfinderapp/data/repository/JobRepository.kt
package com.deepakmanktala.jobfinderapp.data.repository

import com.deepakmanktala.jobfinderapp.data.api.RetrofitClient
import com.deepakmanktala.jobfinderapp.data.model.Job
import com.deepakmanktala.jobfinderapp.data.model.JobSearchRequest
import com.deepakmanktala.jobfinderapp.data.model.PagedJobs

sealed class Result<out T> {
    data class Success<T>(val data: T) : Result<T>()
    data class Error(val message: String) : Result<Nothing>()
    object Loading : Result<Nothing>()
}

class JobRepository {

    private val api = RetrofitClient.instance

    suspend fun searchJobs(request: JobSearchRequest): Result<List<Job>> {
        return try {
            val response = api.searchJobs(request)
            if (response.isSuccessful) {
                Result.Success(response.body()?.jobs ?: emptyList())
            } else {
                Result.Error("Server error: ${response.code()}")
            }
        } catch (e: Exception) {
            Result.Error(e.message ?: "Network error")
        }
    }

    suspend fun getJobs(page: Int = 0, size: Int = 20,
                        keyword: String? = null, region: String? = null): Result<PagedJobs> {
        return try {
            val response = api.getJobs(page, size, keyword, region)
            if (response.isSuccessful) {
                Result.Success(response.body()!!)
            } else {
                Result.Error("Server error: ${response.code()}")
            }
        } catch (e: Exception) {
            Result.Error(e.message ?: "Network error")
        }
    }

    suspend fun getRegions(): Result<List<String>> {
        return try {
            val response = api.getRegions()
            if (response.isSuccessful) Result.Success(response.body()!!)
            else Result.Error("Server error: ${response.code()}")
        } catch (e: Exception) {
            Result.Error(e.message ?: "Network error")
        }
    }
}
